squirrel_control/src/squirrel_hw_interface.cpp

```diff
@@ -0,0 +1,508 @@
+	nh.getParam(param_name, urdf_string);
+      }
+      
+      usleep(100000);
+    }
+    
+    if (!urdf_model_->initString(urdf_string))
+      ROS_ERROR_STREAM_NAMED(name_, "Unable to load URDF model");
+    else
+      ROS_DEBUG_STREAM_NAMED(name_, "Received URDF from param server");
+  }
+  
+
+  void SquirrelHWInterface::doSwitch(const std::list<hardware_interface::ControllerInfo> &start_list, 
+				     const std::list<hardware_interface::ControllerInfo> &stop_list)
+  {
+    std::list<hardware_interface::ControllerInfo>::const_iterator it;
+    for (it = stop_list.begin(); it != stop_list.end(); ++it)
+      {
+	if (it->hardware_interface != JOINT_STATE_INTERFACE) {
+	  enabled_ = false;
+	}
+      }
+    
+    hardware_interface::RobotHW::doSwitch(start_list, stop_list);
+    
+    for (it = start_list.begin(); it != start_list.end(); ++it) {
+      if (it->hardware_interface != JOINT_STATE_INTERFACE) {
+	if (it->hardware_interface == POSITION_JOINT_INTERFACE)	{
+	  current_mode_ = control_modes::POSITION_MODE;
+	  enabled_ = true;
+	}
+	else if (it->hardware_interface == VELOCITY_JOINT_INTERFACE) {
+	  current_mode_ = control_modes::VELOCITY_MODE;
+	  enabled_ = true;
+	}
+	else if (it->hardware_interface == EFFORT_JOINT_INTERFACE) {
+	  current_mode_ = control_modes::TORQUE_MODE;
+	  enabled_ = true;
+	}
+	else {
+	  throw_control_error(true, "Unknown control mode " << it->hardware_interface);
+	}
+      }
+    }
+    motor_interface_->setMode(current_mode_);
+  }
+
+
+  void SquirrelHWInterface::read(ros::Duration &elapsed_time) {
+    switch(current_mode_) {
+    case control_modes::POSITION_MODE:
+      {
+	joint_position_[0] = posBuffer_[0];
+	joint_position_[1] = posBuffer_[1];
+	joint_position_[2] = posBuffer_[2];
+	auto positions = motor_interface_->read();
+	for (int i = 0; i < num_joints_-3; i++) {
+	  joint_position_[i + 3] = positions[i];
+	}
+      }
+      break;
+    case control_modes::VELOCITY_MODE:
+      {
+	joint_velocity_[0] = velBuffer_[0];
+	joint_velocity_[1] = velBuffer_[1];
+	joint_velocity_[2] = velBuffer_[2];
+	auto velocities = motor_interface_->read();
+	for (int i = 0; i < num_joints_-3; i++) {
+	  joint_velocity_[i + 3] = velocities[i];
+	}
+      }
+      break;
+    case control_modes::TORQUE_MODE:
+      {
+	//TODO: assumption: no clue - check base torques
+	joint_effort_[0] = velBuffer_[0];
+	joint_effort_[1] = velBuffer_[1];
+	joint_effort_[2] = velBuffer_[2];
+	auto torques = motor_interface_->read();
+	for (int i = 0; i < num_joints_-3; i++) {
+	  joint_effort_[i + 3] = torques[i];
+	}
+      }
+      break;
+    default:
+      throw_control_error(true, "Unknown mode: " << current_mode_);
+    }
+  }
+
+
+  void SquirrelHWInterface::write(ros::Duration &elapsed_time) {
+    //This for that the robot keeps its initial pose and does not move back to 0 (deault initialization of C++ for class members)
+    if (hold) {
+      joint_position_command_ = joint_position_;
```

Comment:
They are, but just once. When starting the controller and creating the HW interface "hold" is initialized to true. Thus, when executing write() for the first time after being launched the controller uses the current joint values to initialize the joint_position_command_ (as of being a class member C++ initializes all values to 0.0) for that the arm stays in its initial pose - this is really important - imagine the controller launching and crushing the hull as of suddenly moving to its zero position. Anyways, after being executed once, "hold" is set to false and this if is never executed again. Maybe there exists a nicer solution, but for now this works just perfect. This is actually also were we set all base values to 0.0 but the arm just moves.